### ISSUE TITLE
Encryption using parallel stream is slower than default provider

### ISSUE THREAD

**Issue by opensourcegeek** [corretto/amazon-corretto-crypto-provider#118]
I'm trying to run some basic tests to get a feel for perf boost for encryption using KMS with ACCP. I tried running encryption with a JSON string as plaintext with and without ACCP enabled. Following are the numbers I'm seeing, I might have got something wrong - could you please let me know if you could spot something? If I use single thread to run whole 100k I see ACCP being faster but not by much (not even 2x faster). 

I'm on Java 11, 64-bit intel i7 Arch Linux. The numbers are with cache enabled.

10k - encryption - default crypto provider - 23279ms
10k - encryption - amazon corretto crypto provider - 19882ms
Difference 3397ms

100k - encryption - default provider - 222708ms
100k - encryption - amazon corretto crypto provider - 185254ms
Difference 37454ms

Parallel - chunked (100 * 1000)
100k - encryption - default provider - 47032ms
100k - encryption - amazon corretto crypto provider - 48181ms

```java
@Slf4j
@Component
public class KmsUtils {

    @Autowired ApplicationProperties applicationProperties;

    private AwsCrypto awsCrypto;
    private CryptoMaterialsManager dataKeyCache;
    private KmsMasterKeyProvider kmsMasterKeyProvider;

    public String decrypt(String cipherTxt) {
        return awsCrypto.decryptString(dataKeyCache, cipherTxt).getResult();
        // return awsCrypto.decryptString(kmsMasterKeyProvider, cipherTxt).getResult();
    }

    public String encrypt(String clearTxt) {
        return awsCrypto.encryptString(dataKeyCache, clearTxt).getResult();
        // return awsCrypto.encryptString(kmsMasterKeyProvider, clearTxt).getResult();
    }

    @PostConstruct
    void init() {

        log.info("Initialising.");

        awsCrypto = new AwsCrypto();
        awsCrypto.setEncryptionAlgorithm(ALG_AES_256_GCM_IV12_TAG16_HKDF_SHA384_ECDSA_P384);

        kmsMasterKeyProvider =
                KmsMasterKeyProvider.builder()
                        .withDefaultRegion(EU_WEST_2.getName())
                        .withCredentials(new DefaultAWSCredentialsProviderChain())
                        .withKeysForEncryption(applicationProperties.getKmsKeyArn())
                        .build();

        CryptoMaterialsCache cache =
                new LocalCryptoMaterialsCache(applicationProperties.getDataKeyCacheCapacity());

        dataKeyCache =
                CachingCryptoMaterialsManager.newBuilder()
                        .withMasterKeyProvider(kmsMasterKeyProvider)
                        .withCache(cache)
                        .withMaxAge(
                                applicationProperties.getDataKeyCacheMaxEntryAge(),
                                TimeUnit.SECONDS)
                        .withMessageUseLimit(applicationProperties.getDataKeyCacheMaxPerKey())
                        .build();
    }
}


```

```java
    @Test
    public void testKmsUtils_MassiveEncryption() {
        AmazonCorrettoCryptoProvider.install();
        AmazonCorrettoCryptoProvider.INSTANCE.assertHealthy();
        List<Pair<Integer, Integer>> allRanges = new ArrayList<>();

        for (int i = 0; i < 100; i++) {
            int current = i * 1000;
            allRanges.add(Pair.of(current, current + 1000));
        }

        log.info("All pairs {}", allRanges);

        long start = System.currentTimeMillis();
        allRanges.stream()
                .parallel()
                .forEach(
                        tuple -> {
                            int countStart = tuple.getLeft();
                            int countEnd = tuple.getRight();
                            for (int i = countStart; i < countEnd; i++) {
                                String encrypted = kmsutils.encrypt(massiveList);
                            }
                        });

        long finish = System.currentTimeMillis();
        log.info("Total time took for encryption is {} ms", finish - start);
    }
```

**Comment by SalusaSecondus** [comment]
We're looking into this. Thank you.

**Comment by SalusaSecondus** [comment]
Thank you. I've been able to produce a minimal reproduction of this case and agree that in certain circumstances ECDSA signature over NIST P-384 are slower in ACCP than with the default provider.

The root cause of this appears to be the cost of Openssl [continually recalculating an unneeded EC Public key](https://github.com/openssl/openssl/blob/f844f9eb44186df2f8b0cfd3264b4eb003d8c61a/crypto/ec/ec_asn1.c#L1044) when generating signatures.

```
Java_com_amazon_corretto_crypto_provider_EvpSignatureRaw_signRaw
    EVP_PKCS82PKEY
        eckey_priv_decode
            d2i_ECPrivateKey
                EC_POINT_mul (likely called by keygenpub from prior link)
```

Fixing this will likely require significant effort and rearchitecture of ACCP as it will necessitate implementing an EC KeyFactory and custom implementations for Keys.  This will go on our backlog.

**Comment by opensourcegeek** [comment]
Thanks for looking into it. Is there any way to work around this issue, by possibly using a different algorithm? Also I didn't see a massive difference in performance even in single threaded approach, is that also related to the same underlying issue?

**Comment by SalusaSecondus** [comment]
Yes, this is all related to the same issue.

If you do not need an algorithm suite which provided ECDSA signing, then you can select on of them instead and will see significantly improved performance.

[Supported algorithm suites in the AWS Encryption SDK](https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/supported-algorithms.html)

**Comment by opensourcegeek** [comment]
Thanks, I've ran another test using an algorithm without ECDSA and increased the number of cycles to 1M as 100K was too small a sample size for this algorithm,

```java     awsCrypto.setEncryptionAlgorithm(CryptoAlgorithm.ALG_AES_256_GCM_IV12_TAG16_HKDF_SHA256);
```

Below are the numbers, default provider is still not massively behind ACCP. 

Without ECDSA
1M - encryption - default provider - 27162ms
1M - encryption - ACCP - 26211ms

Any ideas on why I'm not seeing bigger difference between default provider and ACCP please?

**Comment by SalusaSecondus** [comment]
I ran a minimal test with the following changes:
1. Used an AES key rather than KMS to avoid measuring network latency and calls
2. set a cache capacity of 1000 keys with 1000 seconds and a message use limit of 100 (since it wasn't specified in your example).
3. JMH for actual benchmarking
4. Didn't use JSON, rather just the `toString` output of your `allRanges` variable.

I'm seeing between 30%-35% improved throughput by using ACCP in this case.

Single threaded:
```
Benchmark               (useAccp)   Mode  Cnt      Score     Error  Units
MyBenchmark.testMethod      false  thrpt    5  20518.644 ± 165.640  ops/s
MyBenchmark.testMethod       true  thrpt    5  26955.313 ± 352.418  ops/s
```

Sixteen threads:
```
Benchmark               (useAccp)   Mode  Cnt      Score     Error  Units
MyBenchmark.testMethod      false  thrpt    5  42786.614 ± 778.858  ops/s
MyBenchmark.testMethod       true  thrpt    5  58022.610 ± 877.810  ops/s
```

Can you please try without using KMS? I'm guessing there is something related to your network calls which is slowing things down.

**Comment by opensourcegeek** [comment]
Yea sure, I'll try it tonight and come back with the numbers. I wasn't sure about KMS as well, I didn't know if there were too many network calls. I'll try and capture some outgoing traffic as well to see how often the calls are made. Thanks again for taking time to look into this.

**Comment by SalusaSecondus** [comment]
@opensourcegeek Just checking to see if you collected any numbers.

Also, I'm wondering if you'd be willing to try the **non-production and for testing purposes only** branch of ACCP? I'd be interested to know how the performance looks to you.
https://github.com/SalusaSecondus/amazon-corretto-crypto-provider/tree/evp_factory

**Comment by opensourcegeek** [comment]
@SalusaSecondus - I haven't had a chance to try plain AES without any KMS yet. However, I did try tweaking KMS cache and seeing the performance after that. The numbers weren't still far off. There was one get call to get the key but after that there were no network calls. Also, checking performance for crypto work is in current sprint so I'll hopefully be picking it up by Monday.

In terms of trying the non-prod version, yea definitely. I'd love to give that a go and see if it improves our performance. Thanks.

**Comment by opensourcegeek** [comment]
Apologies for not getting back earlier @SalusaSecondus - we've decided to stick with default provider for reasons other than performance too. Also, the performance we're seeing with default provider isn't far off from our expectation. So, it's good enough for now. I'll give your non-prod version a try later this evening perhaps, just to see the difference.

**Comment by SalusaSecondus** [comment]
Closing as we've identified the areas for improvement and the specific bottlenecks (ECDSA) for the caller.

**Comment by SalusaSecondus** [comment]
This is fixed by #132
